File: Leetcode/src/com/leetcode/stackPriorityQueue/_224_BasicCalculator.java
package com.leetcode.stackPriorityQueue;

import java.util.Stack;

public class _224_BasicCalculator {

    /**
     *  224. Basic Calculator
        When: 2019/06/15

        solution: 如下所示
     * @param s
     * @return
     */
    public int calculate(String s) {
        // 使用一个stack来实现，之前那种方法对于后缀表达式很有用，但是这种情况还是使用如下的方式
        Stack<Integer> stack = new Stack<Integer>();
        int res = 0;
        int number = 0;
        int sign = 1; //用来标记正负
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (Character.isDigit(c)) {
                number = 10 * number + (int)(c - '0');
            } else if (c == '+') {
                res += sign * number;
                number = 0;
                sign = 1;
            } else if (c == '-') {
                res += sign * number;
                number = 0;
                sign = -1;
            } else if (c == '(') {
                // 相当于计算每一个括号内的结果 后面就依次向外弹出
                stack.push(res);
                stack.push(sign);
                sign = 1;
                res = 0;
            } else if (c == ')') {
                res += sign * number;
                number = 0;
                res *= stack.pop();
                res += stack.pop();
            }
        }
        if (number != 0) res += sign * number;
        return res;
    }
}

File: Leetcode/src/com/leetcode/stack_priority_queue/Calculator/_224_basic_calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        sign = "+"
        num, n = 0, len(s)
        stack = []

        i = 0
        while i < n:
            ch = s[i]

            if ch.isdigit():
                num = num * 10 + int(ch)
            elif ch == "(":
                count = 0
                j = i
                while i < n:
                    if s[i] == "(":
                        count += 1
                    elif s[i] == ")":
                        count -= 1
                    if count == 0:
                        break
                    i += 1
                # anything between "(" number... ")"
                num = self.calculate(s[j + 1 : i])

            if ch == "+" or ch == "-" or i == n - 1:
                if sign == "+":
                    stack.append(num)
                elif sign == "-":
                    stack.append(-num)

                sign = ch
                num = 0

            i += 1
        return sum(stack)
